src/tests/Loader/classloader/ExtendedLayout/CStruct.cs

```diff
@@ -0,0 +1,101 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Xunit;
+
+namespace ExtendedLayoutTests;
+
+public static class CStructTests
+{
+    [Fact]
+    public static void BlittablePrimitiveFieldsLayout()
+    {
+        var c = default(CStructBlittablePrimitiveFields);
+
+        Assert.Equal(12, Unsafe.SizeOf<CStructBlittablePrimitiveFields>());
+        Assert.Equal(0, Unsafe.ByteOffset(ref Unsafe.As<CStructBlittablePrimitiveFields, byte>(ref c), ref Unsafe.As<int, byte>(ref c.a)));
+        Assert.Equal(4, Unsafe.ByteOffset(ref Unsafe.As<CStructBlittablePrimitiveFields, byte>(ref c), ref Unsafe.As<float, byte>(ref c.b)));
+        Assert.Equal(8, Unsafe.ByteOffset(ref Unsafe.As<CStructBlittablePrimitiveFields, byte>(ref c), ref c.c));
+    }
+
+    [Fact]
+    public static void NonBlittableUnmanagedPrimitiveFields_TreatedAsBlittable()
+    {
+        Assert.Equal(Unsafe.SizeOf<CStructNonBlittablePrimitiveFields>(), Marshal.SizeOf<CStructNonBlittablePrimitiveFields>());
+    }
+
+    [Fact]
+    public static void ReferenceFields_ThrowTypeLoadException()
+    {
+        Assert.Throws<TypeLoadException>(() => typeof(CStructWithReferenceFields));
+
+        Assert.Throws<TypeLoadException>(() => typeof(CStructWithMixedFields));
+    }
+
+    [Fact]
+    public static void NestedCStruct()
+    {
+        var nested = new CStructCustomCStructField
+        {
+            y = new NestedCStructType
+            {
+                x = 42
+            }
+        };
+
+        Assert.Equal(4, Unsafe.SizeOf<CStructCustomCStructField>());
+    }
+
+    [Fact]
+    public static void NestedNonCStructNonAuto()
+    {
+        var nested = new CStructCustomSeqStructField
+        {
+            y = new NestedSequentialType
+            {
+                x = 42
+            }
+        };
+
+        Assert.Equal(4, Unsafe.SizeOf<CStructCustomSeqStructField>());
+    }
+
+    [Fact]
+    public static void NestedAutoLayout_ThrowTypeLoadException()
+    {
+        Assert.Throws<TypeLoadException>(() => typeof(CStructCustomAutoStructField));
+    }
+
+    [Fact]
+    public static void EmptyStruct()
+    {
+        Assert.Equal(1, Unsafe.SizeOf<EmptyCStruct>());
+    }
+
+    [Fact]
+    public static void ExplicitOffsets_Ignored()
+    {
+        CStructWithOffsets c = default;
+
+        Assert.Equal(0, Unsafe.ByteOffset(ref Unsafe.As<CStructWithOffsets, byte>(ref c), ref Unsafe.As<int, byte>(ref c.a)));
+    }
+
+    [Fact]
+    public static void ExplicitSize_Ignored()
+    {
+        Assert.Equal(4, Unsafe.SizeOf<CStructWithSize>());
+    }
+
+    [Fact]
+    public static void Pack_Ignored()
+    {
+        Assert.Equal(8, Unsafe.SizeOf<CStructWithPack>());
```

Comment:
Pack of 1 should yield 5. Are we saying that packed semantics aren't going to be supported for now? I see the test is marked "ignored" so I get it. I'm just curious how we think about this and an explicit CStruct ABI type. Not that packing is a C concept.

Reply:
There will be no way to specify Pack with the new attribute outside of IL (as you can’t specify StructLayout and ExtendedLayout at the same time), so I figured that just not respecting it would be the easiest implementation (and say “use Sequential layout if you want custom packing”).

Reply:
Ah, that makes good sense. Thanks.